proposals/simd-0170-builtin-instruction-cost-and-budget.md

```diff
@@ -0,0 +1,105 @@
+---
+simd: '0170'
+title: Allocate precise builtin instructions budget
+authors:
+  - Tao Zhu (Anza)
+category: Standard
+type: Core
+status: Draft
+created: 2024-08-26
+feature: https://github.com/anza-xyz/agave/issues/2562
+supersedes: 
+superseded-by:
+extends:
+---
+
+## Summary
+
+Builtin instructions always consume a statically defined amount of compute
+units (CUs). Therefore, they should allocate the exact same amount of compute
+budget and count the same amount toward the block limit during the banking
+stage.
+
+## Motivation
+
+Builtin instructions in the SVM consume their static DEFAULT_COMPUTE_UNITS from
+the compute budget during execution. These DEFAULT_COMPUTE_UNITS are also
+counted against block limits during the banking stage. However, historically,
+builtin instructions have been allocated DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT
+units of compute budget. This discrepancy between the allowed consumption and
+the actual usage tracked for block limits has led to several issues that need
+to be addressed.
+
+Allocating DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT instead of the actual
+DEFAULT_COMPUTE_UNITS for builtin instructions distorts the tracking of
+transaction compute budgets. This can result in more expensive instructions
+being executed when they should fail due to exceeding the budget. Consequently,
```

Comment:
Good question, I should word it more clearly. The issue is if block packer reserves `150` cu for `transfer`, but CU meter is allocateed `200_000` for it, the following BPF ix can take additional `199_850` CU without burst CU meter, that'd make block packer to adjust *up* `199_850` CUs. Adjust-up is considered bug.

In case of two BPF ixs, both block packer and CU meter see `400_000` CUs, so even 2nd ix can take advantage of left-over CUs from 1st, there wouldn't be case of adjust-up.

Reply:
Okay gotcha, thanks, that makes sense! So by allocating `ALLOCATE_COMPUTE_UNITS` instead of `200_000`, in the case where `ALLOCATE_COMPUTE_UNITS` > `EXECUTE_COMPUTE_UNITS`, is there still an adjust-up between `ALLOCATE_COMPUTE_UNITS` and `EXECUTE_COMPUTE_UNITS`?

Reply:
In that case, BPF instruction can still consume the left-over CUs (eg, up to `ALLOCATE_COMPUTE_UNITS` - total executed ); but since block producer should reserved `ALLOCATE_COMPUTE_UNITS` for builtin, and default 200K for BPF instruction, which is the same CUs allocated for compute-budget for VM, the most CUs combined instructions can consume is the sum of `ALLOCATE_COMPUTE_UNITS` and `200K`, so there can't be adjust-up.